### PR TITLE
Move char filters into analysis-common

### PR DESCRIPTION
**Pull request by nik9000**
Another step down the road to dropping the
lucene-analyzers-common dependency from core.

Note that this removes some tests that no longer compile from
core. I played around with adding them to the analysis-common
module where they would compile but we already test these in
the tests generated from the example usage in the documentation.

I'm not super happy with the way that `requriesAnalysisSettings`
works with regards to plugins. I think it'd be fairly bug-prone
for plugin authors to use. But I'm making it visible as is for
now and I'll rethink later.

A part of #23658

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/indices/analysis/AnalysisModule.java [hunk 1]
```diff
@@ -340,19 +336,6 @@ public AnalysisRegistry getAnalysisRegistry() {
         return normalizers;
     }
 
-    private static <T> AnalysisModule.AnalysisProvider<T> requriesAnalysisSettings(AnalysisModule.AnalysisProvider<T> provider) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/indices/analysis/AnalysisModule.java]: Move to AnalysisPlugin.

File: core/src/test/java/org/elasticsearch/index/analysis/CharFilterTests.java [hunk 2]
```diff
@@ -1,86 +0,0 @@
...
-import org.elasticsearch.cluster.metadata.IndexMetaData;
-import org.elasticsearch.common.settings.Settings;
-import org.elasticsearch.env.Environment;
-import org.elasticsearch.index.IndexSettings;
-import org.elasticsearch.test.ESTokenStreamTestCase;
-import org.elasticsearch.test.IndexSettingsModule;
-
-import static org.elasticsearch.test.ESTestCase.createTestAnalysis;
-
-public class CharFilterTests extends ESTokenStreamTestCase {
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/index/analysis/CharFilterTests.java]: These are duplicates of the rest tests in analysis-common. I can move this to analysis-common and keep it if anyone feels strongly about it.

File: core/src/test/java/org/elasticsearch/indices/analysis/AnalysisModuleTests.java [hunk 3]
```diff
@@ -90,6 +89,11 @@ public AnalysisRegistry getNewRegistry(Settings settings) {
                 public Map<String, AnalysisProvider<TokenFilterFactory>> getTokenFilters() {
                     return singletonMap("myfilter", MyFilterTokenFilterFactory::new);
                 }
+
+                @Override
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/indices/analysis/AnalysisModuleTests.java]: I believe this is a leftover. I'll look into removing this.

File: core/src/test/java/org/elasticsearch/indices/analysis/AnalysisModuleTests.java [hunk 4]
```diff
@@ -184,29 +188,12 @@ private void testSimpleConfiguration(Settings settings) throws IOException {
...
         // verify position increment gap
         analyzer = indexAnalyzers.get("custom6").analyzer();
         assertThat(analyzer, instanceOf(CustomAnalyzer.class));
         CustomAnalyzer custom6 = (CustomAnalyzer) analyzer;
         assertThat(custom6.getPositionIncrementGap("any_string"), equalTo(256));
 
-        // verify characters  mapping
-        analyzer = indexAnalyzers.get("custom5").analyzer();
-        assertThat(analyzer, instanceOf(CustomAnalyzer.class));
-        CustomAnalyzer custom5 = (CustomAnalyzer) analyzer;
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/indices/analysis/AnalysisModuleTests.java]: I could recreate these with mock car filters if anyone feels strongly about it but I think we have these covered very well in other places.

File: core/src/test/java/org/elasticsearch/indices/analyze/AnalyzeActionIT.java [hunk 5]
```diff
@@ -111,36 +111,6 @@ public void testAnalyzeWithNoIndex() throws Exception {
         assertThat(analyzeResponse.getTokens().get(0).getPositionLength(), equalTo(1));
     }
 
-    public void testAnalyzeWithCharFilters() throws Exception {
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/indices/analyze/AnalyzeActionIT.java]: Covered by the rest tests in analysis-common.

File: core/src/test/java/org/elasticsearch/indices/analyze/AnalyzeActionIT.java [hunk 6]
```diff
@@ -263,46 +233,6 @@ public void testAnalyzerWithMultiValues() throws Exception {
...
-                    .put("index.analysis.analyzer.test_analyzer.tokenizer", "standard")
-                    .putArray("index.analysis.analyzer.test_analyzer.char_filter", "my_mapping")
-                    .putArray("index.analysis.analyzer.test_analyzer.filter", "snowball")));
-        ensureGreen();
-
-        for (int i = 0; i < 10; i++) {
-            AnalyzeResponse analyzeResponse = admin().indices().prepareAnalyze().setIndex(indexOrAlias()).setText("THIS IS A PHISH")
-                .setExplain(true).addCharFilter("my_mapping").setTokenizer("keyword").addTokenFilter("lowercase").get();
-
-            assertThat(analyzeResponse.detail().analyzer(), IsNull.nullValue());
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/indices/analyze/AnalyzeActionIT.java]: Covered by the rest tests in analysis-common.

File: core/src/test/java/org/elasticsearch/indices/analyze/AnalyzeActionIT.java [hunk 7]
```diff
@@ -414,90 +344,6 @@ public void testDetailAnalyzeWithMultiValues() throws Exception {
         assertThat(token.getPositionLength(), equalTo(1));
     }
 
-    public void testDetailAnalyzeWithMultiValuesWithCustomAnalyzer() throws Exception {
-        assertAcked(prepareCreate("test").addAlias(new Alias("alias"))
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/indices/analyze/AnalyzeActionIT.java]: Also covered by the rest tests in analysis-common.

File: core/src/test/java/org/elasticsearch/indices/analyze/AnalyzeActionIT.java [hunk 8]
```diff
@@ -575,35 +421,6 @@ public void testCustomTokenFilterInRequest() throws Exception {
         assertThat(analyzeResponse.detail().tokenfilters()[1].getTokens()[0].getPositionLength(), equalTo(1));
     }
 
-
-    public void testCustomCharFilterInRequest() throws Exception {
-        Map<String, Object> charFilterSettings = new HashMap<>();
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/indices/analyze/AnalyzeActionIT.java]: Same.

**Review by rjernst — APPROVED** [review]
LGTM. There are certainly things I think we should cleanup eventually, like `requiresAnalysisSettings()`, but this change is just moving stuff around so I think it is good for now.

**Comment by nik9000** [comment]
Thanks for reviewing @rjernst! I've merged to master only. No backporting for analysis-common.

**Pull request merged**